frame/utility/src/lib.rs

```diff
@@ -15,48 +15,620 @@
+						+ T::MultisigDepositFactor::get() * threshold.into();
+					T::Currency::reserve(&who, deposit)?;
+					<Multisigs<T>>::insert(&id, call_hash, Multisig {
+						when: Self::timepoint(),
+						deposit,
+						depositor: who.clone(),
+						approvals: vec![who.clone()],
+					});
+					Self::deposit_event(RawEvent::NewMultisig(who, id));
+				} else {
+					Err(Error::<T>::NoApprovalsNeeded)?
+				}
+			}
+			Ok(())
+		}
+
+		/// Cancel a pre-existing, on-going multisig transaction. Any deposit reserved previously
+		/// for this operation will be unreserved on success.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `threshold`: The total number of approvals for this dispatch before it is executed.
+		/// - `other_signatories`: The accounts (other than the sender) who can approve this
+		/// dispatch. May not be empty.
+		/// - `timepoint`: The timepoint (block number and transaction index) of the first approval
+		/// transaction for this dispatch.
+		/// - `call_hash`: The hash of the call to be executed.
+		///
+		/// # <weight>
+		/// - `O(S)`.
+		/// - Up to one balance-reserve or unreserve operation.
+		/// - One passthrough operation, one insert, both `O(S)` where `S` is the number of
+		///   signatories. `S` is capped by `MaxSignatories`, with weight being proportional.
+		/// - One encode & hash, both of complexity `O(S)`.
+		/// - One event.
+		/// - I/O: 1 read `O(S)`, one remove.
+		/// - Storage: removes one item.
+		/// # </weight>
+		#[weight = <SigsLen<T::AccountId, Timepoint<T::BlockNumber>>>::new()]
+		fn cancel_as_multi(origin,
+			threshold: u16,
+			other_signatories: Vec<T::AccountId>,
+			timepoint: Timepoint<T::BlockNumber>,
+			call_hash: [u8; 32],
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(threshold >= 1, Error::<T>::ZeroThreshold);
+			let max_sigs = T::MaxSignatories::get() as usize;
+			ensure!(!other_signatories.is_empty(), Error::<T>::TooFewSignatories);
+			ensure!(other_signatories.len() < max_sigs, Error::<T>::TooManySignatories);
+			let signatories = Self::ensure_sorted_and_insert(other_signatories, who.clone())?;
+
+			let id = Self::multi_account_id(&signatories, threshold);
+
+			let m = <Multisigs<T>>::get(&id, call_hash)
+				.ok_or(Error::<T>::NotFound)?;
+			ensure!(m.when == timepoint, Error::<T>::WrongTimepoint);
+			ensure!(m.depositor == who, Error::<T>::NotOwner);
+
+			let _ = T::Currency::unreserve(&m.depositor, m.deposit);
+			<Multisigs<T>>::remove(&id, call_hash);
+
+			Self::deposit_event(RawEvent::MultisigCancelled(who, timepoint, id));
+			Ok(())
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Derive a sub-account ID from the owner account and the sub-account index.
+	pub fn sub_account_id(who: T::AccountId, index: u16) -> T::AccountId {
+		let entropy = (b"modlpy/utilisuba", who, index).using_encoded(blake2_256);
+		T::AccountId::decode(&mut &entropy[..]).unwrap_or_default()
```

Comment:
it assumes that it's no larger than a hash, but doesn't need to be as big, as the tests show. i think that this is fine. other parts of the codebase also assume that `AccountId`s can be decoded from unformatted data.

Reply:
True. I was more concerned about that `T::AccountId` can be anything (theoretically) and it might be that not all variations of `entropy` are suitable for decoding.

If the other parts do assume that `AccountId` can be decoded from unformatted data then we might want to embrace that and document it explicitly. (Not to suggest doing this in this PR though).